### PR TITLE
DW-5194 - Deploy MinIO service

### PR DESCRIPTION
**Pull request by benjaminsherwood**
Signed-off-by: Benjamin Sherwood <benjaminsherwood@digital.uc.dwp.gov.uk>

### CODE REVIEW

**Review by ChrisScottThomas — DISMISSED** [review]

File: files/minio.service [hunk 1]
```diff
@@ -0,0 +1,31 @@
+[Unit]
+Description=MinIO
+Documentation=https://docs.min.io
+Wants=network-online.target
+After=network-online.target
+AssertFileIsExecutable=/usr/bin/minio
+
+[Service]
+WorkingDirectory=/usr
```
Comment (mattburgess) [inline files/minio.service:9]: To check; I suspect this is where minio stores in-flight uploads, before atomically moving them to the actual bucket directory. If so, it'll have to be a scratch directory under `/opt/minio`

File: files/tarball_ingester_minio.sh [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+S3_MINIO_BINARY="s3://${1}/minio/minio.RELEASE.2020-10-28T08-16-50Z"
+
+echo "Downloading min.io binary..."
+aws s3 cp "$S3_MINIO_BINARY" /usr/bin/minio
+echo "Changing permissions for MinIO"
+chmod +x /usr/bin/minio
+minio --version
+
+echo "Creating user minio"
+useradd minio -m
```
Comment (mattburgess) [inline files/tarball_ingester_minio.sh]: To check; does this create a no-shell user (see https://github.com/minio/minio-service#guidelines).

File: files/tarball_ingester_minio.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+echo "Creating MinIO Volume area"
+mkdir -p /opt/minio
+chown -R minio:minio /opt/minio
+
+echo "Creating MinIO Service File"
```
Comment (mattburgess) [inline files/tarball_ingester_minio.sh]:
```suggestion
echo "Creating MinIO Config File"
```

File: tarball-ingester.tf [hunk 4]
```diff
@@ -382,6 +389,26 @@ resource "aws_security_group_rule" "tarball_ingester_ingress_dks" {
   security_group_id        = data.terraform_remote_state.crypto.outputs.dks_sg_id[local.environment]
 }
 
+resource "aws_security_group_rule" "dataset_generator_to_vpc_endpoints" {
```
Comment (mattburgess) [inline tarball-ingester.tf]:
```suggestion
resource "aws_security_group_rule" "tarball_ingester_to_vpc_endpoints" {
```

File: tarball-ingester.tf [hunk 5]
```diff
@@ -382,6 +389,26 @@ resource "aws_security_group_rule" "tarball_ingester_ingress_dks" {
...
+  from_port                = 443
+  protocol                 = "tcp"
+  security_group_id        = aws_security_group.tarball_ingester.id
+  to_port                  = 443
+  type                     = "egress"
+  source_security_group_id = data.terraform_remote_state.ingest.outputs.vpc.vpc.interface_vpce_sg_id
+}
+
+resource "aws_security_group_rule" "vpc_endpoints_from_dataset_generator" {
+  description              = "Allow HTTPS traffic from Dataset Generator"
```
Comment (mattburgess) [inline tarball-ingester.tf]:
```suggestion
resource "aws_security_group_rule" "vpc_endpoints_from_tarball_ingester" {
  description              = "Allow HTTPS traffic from Tarball Ingester"
```

File: files/minio.service (hunk 1 above)
Comment (benjaminsherwood) [reply files/minio.service:9]: no, WorkingDirectory is a systemd construct, it's just the context that Exec gets done from, and we're defining that as an explicit path anyways so it doesn't matter.

File: files/tarball_ingester_minio.sh (hunk 2 above)
Comment (benjaminsherwood) [reply files/tarball_ingester_minio.sh]: was copy-pasted and should in fact be a capital M, an no that's not no-shell, replaced with:

useradd minio -M --shell=/bin/false

File: tarball-ingester.tf [hunk 6]
```diff
@@ -382,6 +389,26 @@ resource "aws_security_group_rule" "tarball_ingester_ingress_dks" {
   security_group_id        = data.terraform_remote_state.crypto.outputs.dks_sg_id[local.environment]
 }
 
+resource "aws_security_group_rule" "tarball_ingester_to_vpc_endpoints_to_vpc_endpoints" {
```
Comment (mattburgess) [inline tarball-ingester.tf]:
```suggestion
resource "aws_security_group_rule" "tarball_ingester_to_vpc_endpoints" {
```

File: tarball-ingester.tf [hunk 7]
```diff
@@ -382,6 +389,26 @@ resource "aws_security_group_rule" "tarball_ingester_ingress_dks" {
...
+  description              = "Allow HTTPS traffic to VPC endpoints"
+  from_port                = 443
+  protocol                 = "tcp"
+  security_group_id        = aws_security_group.tarball_ingester.id
+  to_port                  = 443
+  type                     = "egress"
+  source_security_group_id = data.terraform_remote_state.ingest.outputs.vpc.vpc.interface_vpce_sg_id
+}
+
+resource "aws_security_group_rule" "vpc_endpoints_from_tarball_ingester_to_vpc_endpoints" {
```
Comment (mattburgess) [inline tarball-ingester.tf]:
```suggestion
resource "aws_security_group_rule" "vpc_endpoints_from_tarball_ingester" {
```

File: files/tarball_ingester_minio.sh [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+S3_MINIO_BINARY="s3://${1}/minio/minio.RELEASE.2020-10-28T08-16-50Z"
+
+echo "Downloading min.io binary..."
+aws s3 cp "$S3_MINIO_BINARY" /usr/bin/minio
+echo "Changing permissions for MinIO"
+chmod +x /usr/bin/minio
+minio --version
+
+echo "Creating user minio"
+useradd minio -M --shell=/bin/false
```
Comment (mattburgess) [inline files/tarball_ingester_minio.sh]:
```suggestion
useradd minio -M --shell=/sbin/nologin
```

**Review by mattburgess — APPROVED** [review]

**Pull request merged**